### DISCUSSION TITLE
VMESS+Trojan PacketTunnel exceeds 15M limit

### DISCUSSION THREAD

**Discussion by gaoxiaowei** [eycorsican/leaf#108] [Q&A]
I used the Trojan protocol + Speedtest APP to measure the speed, which exceeded the 15M limit after running for a period of time. This is an inevitable problem. Is there an optimization method for this problem? Thank you!

![image](https://user-images.githubusercontent.com/6646699/113652663-c9e74500-96c6-11eb-97b0-cad223852bc0.png)

**Comment by gaoxiaowei** [comment]
H2+VMESS  also have the same problem

![image](https://user-images.githubusercontent.com/6646699/113654994-a07ce800-96cb-11eb-9f5a-4d6bd539a80c.png)

**Reply by tc2024** [reply]
Which version supports vmess protocol？The current version（v0.3.1） does not support vmess protocol

**Comment by eycorsican** [comment]
Logs? Full config?

**Reply by gaoxiaowei** [reply]
h2 config:
{
 "log": {
        "level": "trace"
    },
 "dns": {
        "servers": [
            "1.1.1.1",
            "114.114.114.114"
        ]
    },
  "inbounds": [
      {
          "protocol": "socks",
          "address": "127.0.0.1",
          "port": 1086
      },
      {
        "address" : "127.0.0.1",
        "protocol" : "http",
        "port" : 8001
      },
      {
          "protocol": "tun",
          "settings": {
              "fd": REPLACE-ME-WITH-THE-FD
          },
          "tag": "tun"
      }

  ],
  "outbounds": [
      {
          "protocol": "chain",
          "settings": {
              "actors": [
                  "vmess_tls",
                  "vmess_h2",
                  "vmess"
              ]
          },
          "tag": "vmess_out"
      },
      {
          "protocol": "tls",
          "settings": {
             "serverName": "2783482488.testnet.co",
             "alpn": ["h2"],
             "insecure":true
          },
          "tag": "vmess_tls"
      },
      {
          "protocol": "h2",
          "settings": {
              "host": "2783482488.testnet.co",
              "path": "/try"
          },
          "tag": "vmess_h2"
      },
      {
          "protocol": "vmess",
          "settings": {
              "address": "2783482488.testnet.co",
              "port": 443,
              "uuid": "00725c4a-11d1-42bb-9167-ccee9d79b969"
          },
          "tag": "vmess"
      }
  ]
}

log:

Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][DEBUG] tcp uplink 192.0.2.1:50591 -> paloalto.ca.speedtest.frontier.com:8080 error: read on broken pipe [vmess_out]
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] applied 4s downlink timeout to 192.0.2.1:50591 <- paloalto.ca.speedtest.frontier.com:8080
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] tcp eof 192.0.2.1:58277
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][DEBUG] tcp uplink 192.0.2.1:58277 -> fastlane.rubiconproject.com:443 done, 1429 bytes transfered [vmess_out]
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] applied 4s downlink timeout to 192.0.2.1:58277 <- fastlane.rubiconproject.com:443
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] tcp new 192.0.2.1:58279
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] pick route failed: no matching rules
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][DEBUG] picked default route [vmess_out] for 192.0.2.1:58279 -> fastlane.rubiconproject.com:443
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] active proxied tcp connections +1: 40
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] dialing tcp 165.232.146.120:443
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] tcp new 192.0.2.1:58280
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] pick route failed: no matching rules
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][DEBUG] picked default route [vmess_out] for 192.0.2.1:58280 -> aas.dc-1.net:443
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] active proxied tcp connections +1: 41
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][TRACE] dialing tcp 165.232.146.120:443
Apr  6 11:53:22 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:22][INFO] [tcp] [tun] [vmess_out] [1048ms] aax-us-east.amazon-adsystem.com:443
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] connected tcp 165.232.146.120:443
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] wrapping tls with name 2783482488.acsnet.co
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] stream rustls-tls:wrap_tls option true
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] connected tcp 165.232.146.120:443
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] wrapping tls with name 2783482488.acsnet.co
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] stream rustls-tls:wrap_tls option true
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][INFO] [tcp] [tun] [vmess_out] [1042ms] paloalto.ca.speedtest.frontier.com:8080
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][INFO] [tcp] [tun] [vmess_out] [1040ms] paloalto.ca.speedtest.frontier.com:8080
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][INFO] [tcp] [tun] [vmess_out] [1045ms] paloalto.ca.speedtest.frontier.com:8080
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][INFO] [tcp] [tun] [vmess_out] [1044ms] paloalto.ca.speedtest.frontier.com:8080
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][INFO] [tcp] [tun] [vmess_out] [1044ms] paloalto.ca.speedtest.frontier.com:8080
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][DEBUG] tcp downlink 192.0.2.1:58277 <- fastlane.rubiconproject.com:443 done, 5159 bytes transfered [vmess_out]
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] active proxied tcp connections -1: 40
Apr  6 11:53:23 PacketTunnel[4234] <Notice>: [2021-04-06 11:53:23][TRACE] tcp drop 192.0.2.1:58277

**Reply by eycorsican** [reply]
oom issues on ios usually caused by having too many tls connections, but I don't think ~40 tls connections could cause an oom issue. Perhaps you could further  lower [the limit](https://github.com/eycorsican/leaf/blob/021391608cdb6d5ef16dfb1146f10daf61423885/leaf/src/option/ios.rs#L2) to verify whether this  is the root cause.

**Reply by gaoxiaowei** [reply]
I tried to limit the number of connections but the problem remained. Is it possible that the problem is caused by high instantaneous memory

Below is the maximum number of connections used in my tests:

/// Maximum number of proxy outbound TCP connections allowed at the same time.
pub static ENDPOINT_TCP_CONCURRENCY: usize = 30;

/// Maximum number of direct outbound TCP connections allowed at the same time.
pub static DIRECT_TCP_CONCURRENCY: usize = 34;

/// DNS cache size in the built-in DNS client.
pub static DNS_CACHE_SIZE: usize = 128;

**Reply by eycorsican** [reply]
Can you provide a full log?

**Reply by gaoxiaowei** [reply]
[crash.log](https://github.com/eycorsican/leaf/files/6262828/crash.log)

**Reply by ghost** [reply]
> oom issues on ios usually caused by having too many tls connections, but I don't think ~40 tls connections could cause an oom issue. Perhaps you could further lower [the limit](https://github.com/eycorsican/leaf/blob/021391608cdb6d5ef16dfb1146f10daf61423885/leaf/src/option/ios.rs#L2) to verify whether this is the root cause.

In fact,40 connections is not enough ,it will stuck just like log below:

```
Jan 24 21:41:01 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:01][INFO] [tcp] [Trojan] [1071ms] k.sinaimg.cn:443
Jan 24 21:41:01 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:01][INFO] [tcp] [Trojan] [1063ms] mixern.sina.cn:443
Jan 24 21:41:01 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:01][INFO] [tcp] [Trojan] [1074ms] mixern.sina.cn:443
Jan 24 21:41:01 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:01][INFO] [tcp] [Trojan] [1076ms] mjs.sinaimg.cn:443
Jan 24 21:41:02 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:02][INFO] [tcp] [Trojan] [1710ms] mixern.sina.cn:443
Jan 24 21:41:03 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:03][INFO] [tcp] [Trojan] [3342ms] mixern.sina.cn:443
Jan 24 21:41:08 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:08][INFO] [tcp] [Trojan] [7427ms] interface.sina.cn:443
Jan 24 21:41:12 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:12][INFO] [tcp] [Trojan] [11441ms] beacon.sina.com.cn:443
Jan 24 21:41:16 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:16][INFO] [tcp] [Trojan] [15457ms] 9ibbcq2d.sina.cn:443
Jan 24 21:41:19 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:19][INFO] [tcp] [Trojan] [18366ms] qgnu18wo.sina.cn:443
Jan 24 21:41:19 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:19][INFO] [tcp] [Trojan] [19055ms] beacon.sina.com.cn:443
Jan 24 21:41:19 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:19][INFO] [tcp] [Trojan] [19058ms] interface.sina.cn:443
Jan 24 21:41:20 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:20][INFO] [tcp] [Trojan] [18751ms] courier.push.apple.com:5223
Jan 24 21:41:22 PacketTunnel[10412] <Notice>: [2021-01-24 21:41:22][INFO] [tcp] [Trojan] [19624ms] p33-keyvalueservice.icloud.com:443
```


If U want make it smooth and faster,connection limited num at least must more than **100+**,I do spend lot of time testing.

But that will OOM,absolutely.

**Reply by ghost** [reply]
感觉暂时没有很好的解决办法。15M限制太恶心。

**Reply by gaoxiaowei** [reply]
> 感觉暂时没有很好的解决办法。15M限制太恶心。

Yeah, but if it's not optimized, it's always a time bomb.

**Reply by eycorsican** [reply]
You  may try the `dev` branch and building with the `default-openssl` feature, openssl may not as efficient as ring/rustls, but you got more tunable options.

**Reply by gaoxiaowei** [reply]
I have tested with the dev branch and there is no OOM issue with Trojan. H2 is not only slow, but also has OOM problem. The relevant logs are as follows

[dev_h2_crash.log](https://github.com/eycorsican/leaf/files/6276890/dev_h2_crash.log)

Currently, using OpenSSL has a smaller memory footprint than RustTLS

By the way, using the master branch WS_TLS also causes OOM issues multiple times, which is relatively rare

thanks

**Reply by eycorsican** [reply]
tbh, I've never tried h2 in practice, I don't even know whether there are tunable parameters in  h2.

**Reply by ghost** [reply]
> You may try the `dev` branch and building with the `default-openssl` feature, openssl may not as efficient as ring/rustls, but you got more tunable options.

I tried as U say,openssl has better performance then rustls-tls,really!
maybe [tls_performance_rustls_versus_openssl](https://www.reddit.com/r/programming/comments/c8pymx/tls_performance_rustls_versus_openssl/) is not right ?Or just because U used   [ 'local cert'](https://curl.se/docs/caextract.html) ?

> rustls is 15% quicker to send data.
> rustls is 5% quicker to receive data.
> rustls is 20-40% quicker to set up a client connection.
> rustls is 10% quicker to set up a server connection.
> rustls is 30-70% quicker to resume a client connection.
> rustls is 10-20% quicker to resume a server connection.
> rustls uses less than half the memory of OpenSSL.

**Reply by ghost** [reply]
> > You may try the `dev` branch and building with the `default-openssl` feature, openssl may not as efficient as ring/rustls, but you got more tunable options.
> 
> 
> 
> I tried as U say,openssl has better performance then rustls-tls,really!
> 
> maybe [tls_performance_rustls_versus_openssl](https://www.reddit.com/r/programming/comments/c8pymx/tls_performance_rustls_versus_openssl/) is not right ?Or just because U used   [ 'local cert'](https://curl.se/docs/caextract.html) ?
> 
> 
> 
> > rustls is 15% quicker to send data.
> 
> > rustls is 5% quicker to receive data.
> 
> > rustls is 20-40% quicker to set up a client connection.
> 
> > rustls is 10% quicker to set up a server connection.
> 
> > rustls is 30-70% quicker to resume a client connection.
> 
> > rustls is 10-20% quicker to resume a server connection.
> 
> > rustls uses less than half the memory of OpenSSL.

老哥。最近我发现访问一些网站内存释放不是很及时。你说会不会和内存释放不及时有关
